Relax type hints to DateTimeInterface and make use of DateTimeImmutable

gws opened the pull request:
This change breaks backward compatibility, in that getStart() and getEnd() will always return instances of DateTimeImmutable instead of DateTime as they did previously.

PHP >=5.6 is now required. PHP 5.5 is EOL, and users who cannot upgrade should use older releases of this library.

gws committed:
Relax type hints to DateTimeInterface and make use of DateTimeImmutable

This change breaks backward compatibility, in that getStart() and
getEnd() will always return instances of DateTimeImmutable instead of
DateTime as they did previously.

PHP >=5.6 is now required. PHP 5.5 is EOL, and users who cannot upgrade
should use older releases of this library.

rquadling commented:
I've got some commits on this. Includes coverage to close to 100%. Also upgraded to PHPUnit 6.

PHPStan and php-cs.

I'll create a PR when I get back to work.

rquadling commented:
I'd go with a minor release due to the BC break.

Also considering PHP5.6 is now locked for security fixes only, move to PHP 7+ only.

rquadling commented on library/Vo/DateTimeRange.php:
```diff
@@ -64,14 +64,14 @@ public function diff(DateRange $arg)
             return new static(
-                clone $this->getStart(),
-                date_modify(clone $arg->getStart(), '-1 second')
+                $this->getStart(),
```
This one caught me when I did this. Can't use date_modify() on immutables.

rquadling commented on tests/VoTest/DateRangeTest.php:
```diff
@@ -3,6 +3,7 @@
 namespace VoTest;
 
 use DateTime;
+use DateTimeImmutable;
```
Mutable dates still need to be tested as it is the VO's job to make them immutable.

gws replied:
Immutable dates also need to be tested. There's a mix of mutable and immutable if you look at the tests, though it should be sufficient to test that the constructor, separately, provides an immutable date invariant.

rquadling replied:
Ah. Yes. I missed that.

rquadling commented:
Would you consider PHP7+.

As 5.6 is now feature frozen (http://php.net/supported-versions.php).

rquadling commented:
If so, we could pull in PHPUnit 6 and PHPStan - both PHP7 only.

gws commented:
Since we're making a breaking change requiring a major version release anyway, I'm fine with that. I'll make the changes soon.

gws committed:
Require use of PHP >=7.0 and add PHPStan

Numerous fixes to accommodate changes in PHPUnit as well as some fixes
found by PHPStan.

gws commented:
I believe I've made the necessary changes to pull in the new versions and projects you wanted.
